daemon/vpn/quench/libquench.go

```diff
@@ -0,0 +1,272 @@
+//go:build quench
+
+package quench
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"sync"
+
+	quenchBindigns "quench"
+
+	"github.com/NordSecurity/nordvpn-linux/daemon/vpn"
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	"github.com/NordSecurity/nordvpn-linux/tunnel"
+)
+
+const (
+	vnicName            = "qtun"
+	quenchPrefix        = "[quench]"
+	quenchInterfaceAddr = "10.3.0.2/16"
+)
+
+type Logger struct{}
+
+func (l *Logger) Log(logLevel quenchBindigns.LogLevel, message string) {
+	log.Println(internal.DebugPrefix, quenchPrefix, message)
+}
+
+type Observer struct {
+	mu                        sync.Mutex
+	currentState              vpn.State
+	eventsChan                chan<- vpn.State
+	eventsSubscribtionContext context.Context
+}
+
+func NewObserver() *Observer {
+	return &Observer{
+		eventsChan: nil,
+	}
+}
+
+func (o *Observer) SubscribeToEvents(ctx context.Context) <-chan vpn.State {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	eventsChan := make(chan vpn.State)
+	o.eventsChan = eventsChan
+	o.eventsSubscribtionContext = ctx
+	return eventsChan
+}
+
+func (o *Observer) notifyConnectionStateChange(state vpn.State) {
+	o.currentState = state
+	if o.eventsChan != nil {
+		log.Println(internal.DebugPrefix, quenchPrefix, "unsubscribing from quench state changes")
+		select {
+		case o.eventsChan <- state:
+		case <-o.eventsSubscribtionContext.Done():
+			o.eventsChan = nil
+		}
+	}
+}
+
+func (o *Observer) Connecting() {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	// Log only when state has changet to ConnectingState from some other state. This will prevent log flood when
+	// libquench attempts to reconnect multiple times in no-net scenario.
+	if o.currentState != vpn.ConnectingState {
+		log.Println(internal.DebugPrefix, quenchPrefix, "connecting to quench server")
+	}
+
+	o.notifyConnectionStateChange(vpn.ConnectingState)
+}
+
+func (o *Observer) Connected() {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.notifyConnectionStateChange(vpn.ConnectedState)
+
+	log.Println(internal.DebugPrefix, "connected")
+}
+
+func (o *Observer) Disconnected(reason quenchBindigns.DisconnectReason) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.notifyConnectionStateChange(vpn.ExitedState)
+
+	log.Println(internal.DebugPrefix, quenchPrefix, "disconnected:", reason)
+}
+
+type Quench struct {
+	mu       sync.Mutex
+	fwmark   uint32
+	vnicName string
+	observer *Observer
+	logger   *Logger
+	state    vpn.State
+	server   vpn.ServerData
+	vnic     *quenchBindigns.Vnic
+	tun      *tunnel.Tunnel
+}
+
+func New(fwmark uint32) *Quench {
+	logger := Logger{}
+	quenchBindigns.SetLogCallback(quenchBindigns.LogLevelDebug, &logger)
```

Comment:
Might be good to have different levels of debug based on the build type, and for prod to be set to `info`.